File: sdk/android/src/main/java/dev/dazzle/sdk/edge/OpenAICompatibleClient.kt
// Copyright 2026 Ivan Aliaga <ivan.aliaga@urp.edu.pe>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.

package dev.dazzle.sdk.edge

import dev.dazzle.sdk.Completion
import dev.dazzle.sdk.Delta
import dev.dazzle.sdk.LLMClient
import dev.dazzle.sdk.Message
import dev.dazzle.sdk.Role
import dev.dazzle.sdk.ToolCall
import dev.dazzle.sdk.ToolDeclaration
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL

/**
 * [LLMClient] that speaks the OpenAI `chat/completions` wire format.
 *
 * Any host that serves `/v1/chat/completions` with the OpenAI schema
 * works: OpenAI itself, Azure OpenAI, Groq, Together AI, HuggingFace
 * Inference Providers (`router.huggingface.co/v1`), Ollama local
 * (`localhost:11434/v1`), vLLM, LM Studio, an OpenRouter proxy, or
 * any FastAPI you write yourself.
 *
 * ## Examples
 *
 * ```kotlin
 * // OpenAI
 * val openai = OpenAICompatibleClient(
 *     baseURL = "https://api.openai.com/v1",
 *     model   = "gpt-4o-mini",
 *     apiKey  = BuildConfig.OPENAI_API_KEY,
 * )
 *
 * // HuggingFace Inference (any HF-hosted model)
 * val hf = OpenAICompatibleClient(
 *     baseURL = "https://router.huggingface.co/v1",
 *     model   = "meta-llama/Llama-3.3-70B-Instruct",
 *     apiKey  = BuildConfig.HF_TOKEN,
 * )
 *
 * // Ollama running on the dev machine, reachable from the emulator
 * val ollama = OpenAICompatibleClient(
 *     baseURL = "http://10.0.2.2:11434/v1",
 *     model   = "llama3.2",
 * )
 * ```
 *
 * ## Tool-calling
 *
 * Native wire-format — when the remote reply includes `tool_calls`
 * they are surfaced directly as `Delta.ToolCallStart` +
 * `Delta.ToolCallArgs` (and as `Completion.ToolCalls` from
 * [complete]). No extra parser required.
 *
 * ## Transport
 *
 * Uses `HttpURLConnection` (Android SDK, no external deps). SSE
 * frames are read line-by-line from the response body on
 * [Dispatchers.IO] and parsed with `org.json`.
 *
 * ## Cleartext and ATS
 *
 * For `http://` base URLs (e.g. local Ollama over the emulator
 * bridge `10.0.2.2:11434`), the consumer app must opt in via
 * `android:usesCleartextTraffic="true"` or a
 * `network_security_config.xml` — the SDK does not relax that for
 * you.
 */
class OpenAICompatibleClient(
    private val baseURL: String,
    private val model: String,
    private val apiKey: String? = null,
    private val extraHeaders: Map<String, String> = emptyMap(),
    private val temperature: Double? = null,
    private val maxTokens: Int? = null,
    private val connectTimeoutMs: Int = 30_000,
    private val readTimeoutMs: Int = 120_000,
) : LLMClient {

    override val modelId: String = model

    // ── complete (non-streaming) ─────────────────────────────────────────

    override suspend fun complete(
        messages: List<Message>,
        tools: List<ToolDeclaration>,
    ): Completion = withContext(Dispatchers.IO) {
        val body = encodeBody(messages, tools, stream = false)
        val raw = httpPost(body, sse = false)
        decodeNonStreaming(raw)
    }

    // ── stream (SSE) ─────────────────────────────────────────────────────

    override fun stream(
        messages: List<Message>,
        tools: List<ToolDeclaration>,
    ): Flow<Delta> = flow {
        val body = encodeBody(messages, tools, stream = true)
        val conn = openConnection(body, sse = true)
        try {
            val status = conn.responseCode
            if (status !in 200..299) {
                val errBody = (conn.errorStream ?: conn.inputStream)?.use {
                    BufferedReader(InputStreamReader(it, Charsets.UTF_8)).readText()
                } ?: "<empty>"
                throw OpenAICompatibleException.HttpError(status, errBody)
            }
            // `liveCalls[index] = (id, name)` — tool-call fragments carry
            // an `index` and share an `id` only on the first chunk, so we
            // memoize both so subsequent ToolCallArgs emissions reuse the
            // same id the ChatAgent expects.
            val liveCalls = mutableMapOf<Int, Pair<String, String>>()

            BufferedReader(InputStreamReader(conn.inputStream, Charsets.UTF_8)).use { reader ->
                while (true) {
                    val line = reader.readLine() ?: break
                    if (!line.startsWith("data: ")) continue
                    val payload = line.removePrefix("data: ")
                    if (payload == "[DONE]") break
                    val chunk = JSONObject(payload)
                    val choices = chunk.optJSONArray("choices") ?: continue
                    if (choices.length() == 0) continue
                    val delta = choices.getJSONObject(0).optJSONObject("delta") ?: continue
                    val text = delta.optString("content", "")
                    if (text.isNotEmpty()) emit(Delta.Text(text))
                    val calls = delta.optJSONArray("tool_calls") ?: continue
                    for (i in 0 until calls.length()) {
                        val tc = calls.getJSONObject(i)
                        val index = tc.optInt("index", i)
                        val fn = tc.optJSONObject("function")
                        val name = fn?.optString("name", "")?.takeIf { it.isNotEmpty() }
                        if (name != null) {
                            val id = tc.optString("id", "tc_$index")
                            liveCalls[index] = id to name
                            emit(Delta.ToolCallStart(id = id, name = name))
                        }
                        val args = fn?.optString("arguments", "")?.takeIf { it.isNotEmpty() }
                        if (args != null) {
                            val (id, _) = liveCalls[index] ?: continue
                            emit(Delta.ToolCallArgs(id = id, argsChunk = args))
                        }
                    }
                }
            }
            emit(Delta.End)
        } finally {
            conn.disconnect()
        }
    }.flowOn(Dispatchers.IO)

    override fun close() { /* URL connections are per-call; nothing persistent. */ }

    // ── HTTP helpers ─────────────────────────────────────────────────────

    private fun httpPost(body: ByteArray, sse: Boolean): String {
        val conn = openConnection(body, sse)
        try {
            val status = conn.responseCode
            val stream = if (status in 200..299) conn.inputStream else (conn.errorStream ?: conn.inputStream)
            val text = BufferedReader(InputStreamReader(stream, Charsets.UTF_8)).readText()
            if (status !in 200..299) {
                throw OpenAICompatibleException.HttpError(status, text)
            }
            return text
        } finally {
            conn.disconnect()
        }
    }

    private fun openConnection(body: ByteArray, sse: Boolean): HttpURLConnection {
        val url = URL(baseURL.trimEnd('/') + "/chat/completions")
        val conn = url.openConnection() as HttpURLConnection
        conn.requestMethod = "POST"
        conn.connectTimeout = connectTimeoutMs
        conn.readTimeout = readTimeoutMs
        conn.doOutput = true
        conn.setRequestProperty("Content-Type", "application/json")
        conn.setRequestProperty(
            "Accept",
            if (sse) "text/event-stream" else "application/json",
        )
        if (!apiKey.isNullOrEmpty()) {
            conn.setRequestProperty("Authorization", "Bearer $apiKey")
        }
        for ((k, v) in extraHeaders) conn.setRequestProperty(k, v)
        conn.outputStream.use { it.write(body) }
        return conn
    }

    // ── Wire encode ──────────────────────────────────────────────────────

    private fun encodeBody(
        messages: List<Message>,
        tools: List<ToolDeclaration>,
        stream: Boolean,
    ): ByteArray {
        val body = JSONObject()
        body.put("model", model)
        body.put("stream", stream)
        body.put("messages", JSONArray(messages.map(::wireMessage)))
        temperature?.let { body.put("temperature", it) }
        maxTokens?.let { body.put("max_tokens", it) }
        if (tools.isNotEmpty()) {
            body.put("tools", JSONArray(tools.map(::wireTool)))
            body.put("tool_choice", "auto")
        }
        return body.toString().toByteArray(Charsets.UTF_8)
    }

    /** Map one [Message] to the OpenAI wire shape. */
    private fun wireMessage(m: Message): JSONObject {
        val o = JSONObject()
        o.put("role", m.role.name)
        o.put("content", m.content)
        if (m.toolCalls.isNotEmpty()) {
            val arr = JSONArray()
            for (tc in m.toolCalls) {
                val call = JSONObject()
                call.put("id", tc.id)
                call.put("type", "function")
                val fn = JSONObject()
                fn.put("name", tc.name)
                // arguments is a raw JSON string in the Dazzle model and
                // OpenAI wants a JSON string too — pass as-is.
                fn.put("arguments", tc.arguments)
                call.put("function", fn)
                arr.put(call)
            }
            o.put("tool_calls", arr)
        }
        m.toolCallId?.let { o.put("tool_call_id", it) }
        return o
    }

    private fun wireTool(d: ToolDeclaration): JSONObject {
        val schemaObj = JSONObject(d.parameters.serialize())
        val fn = JSONObject()
        fn.put("name", d.name)
        fn.put("description", d.description)
        fn.put("parameters", schemaObj)
        val tool = JSONObject()
        tool.put("type", "function")
        tool.put("function", fn)
        return tool
    }

    // ── Non-streaming decode ─────────────────────────────────────────────

    private fun decodeNonStreaming(raw: String): Completion {
        val obj = JSONObject(raw)
        val choices = obj.optJSONArray("choices")
            ?: throw OpenAICompatibleException.EmptyResponse
        if (choices.length() == 0) throw OpenAICompatibleException.EmptyResponse
        val message = choices.getJSONObject(0).optJSONObject("message")
            ?: throw OpenAICompatibleException.EmptyResponse
        val toolCallsArr = message.optJSONArray("tool_calls")
        if (toolCallsArr != null && toolCallsArr.length() > 0) {
            val calls = mutableListOf<ToolCall>()
            for (i in 0 until toolCallsArr.length()) {
                val tc = toolCallsArr.getJSONObject(i)
                val fn = tc.optJSONObject("function")
                calls += ToolCall(
                    id        = tc.optString("id", "tc_$i"),
                    name      = fn?.optString("name", "").orEmpty(),
                    arguments = fn?.optString("arguments", "{}") ?: "{}",
                )
            }
            return Completion.ToolCalls(
                Message(role = Role.assistant, content = "", toolCalls = calls)
            )
        }
        return Completion.Text(
            Message(role = Role.assistant, content = message.optString("content", ""))
        )
    }
}

/** Sealed error class for HTTP-level failures from the adapter. */
sealed class OpenAICompatibleException(msg: String) : RuntimeException(msg) {
    class HttpError(val status: Int, val body: String)
        : OpenAICompatibleException("OpenAI-compatible HTTP $status: $body")
    data object EmptyResponse
        : OpenAICompatibleException("OpenAI-compatible response had no choices")
}

File: sdk/ios/Sources/OpenAICompatibleClient.swift
// Copyright 2026 Ivan Aliaga <ivan.aliaga@urp.edu.pe>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.

import Foundation

/// `LLMClient` that speaks the OpenAI `chat/completions` wire format.
///
/// Any host that serves `/v1/chat/completions` with the OpenAI
/// schema works: OpenAI itself, Azure OpenAI, Groq, Together AI,
/// HuggingFace Inference Providers (`router.huggingface.co/v1`),
/// Ollama local (`localhost:11434/v1`), vLLM, LM Studio, an
/// OpenRouter proxy, or any FastAPI you write yourself.
///
/// ## Examples
///
/// ```swift
/// // OpenAI
/// let openai = OpenAICompatibleClient(
///     baseURL: URL(string: "https://api.openai.com/v1")!,
///     model: "gpt-4o-mini",
///     apiKey: ProcessInfo.processInfo.environment["OPENAI_API_KEY"]
/// )
///
/// // HuggingFace Inference (any HF-hosted model)
/// let hf = OpenAICompatibleClient(
///     baseURL: URL(string: "https://router.huggingface.co/v1")!,
///     model: "meta-llama/Llama-3.3-70B-Instruct",
///     apiKey: ProcessInfo.processInfo.environment["HF_TOKEN"]
/// )
///
/// // Ollama running on the dev machine, reachable from the simulator
/// let ollama = OpenAICompatibleClient(
///     baseURL: URL(string: "http://localhost:11434/v1")!,
///     model: "llama3.2"
/// )
/// ```
///
/// Native tool-call emission: when the remote reply includes
/// `tool_calls`, they are surfaced as `Delta.toolCallStart` /
/// `Delta.toolCallArgs` on the stream (and as `Completion.toolCalls`
/// from `complete`). No extra parser needed — the wire format is
/// already structured.
///
/// ## Transport
///
/// Uses `URLSession` directly, no external dependency. Streaming
/// parses SSE (`data: {...}\n\n`) with a hand-rolled buffer reader.
/// Cancellation: the caller's Task cancellation propagates through
/// the `AsyncThrowingStream` and aborts the HTTP body.
///
/// ## Security
///
/// HTTPS is enforced by iOS App Transport Security by default. To
/// reach a plain-HTTP host (e.g. local Ollama on `http://`), the
/// consumer app must opt in via `NSAppTransportSecurity` in its
/// `Info.plist` — the SDK does not disable ATS for you.
public final class OpenAICompatibleClient: LLMClient, @unchecked Sendable {

    public let modelId: String

    private let baseURL: URL
    private let model: String
    private let apiKey: String?
    private let extraHeaders: [String: String]
    private let session: URLSession
    private let temperature: Double?
    private let maxTokens: Int?

    public init(
        baseURL: URL,
        model: String,
        apiKey: String? = nil,
        extraHeaders: [String: String] = [:],
        temperature: Double? = nil,
        maxTokens: Int? = nil,
        session: URLSession = .shared
    ) {
        self.baseURL = baseURL
        self.model = model
        self.apiKey = apiKey
        self.extraHeaders = extraHeaders
        self.session = session
        self.temperature = temperature
        self.maxTokens = maxTokens
        self.modelId = model
    }

    // MARK: – LLMClient

    public func complete(
        messages: [Message],
        tools: [ToolDeclaration]
    ) async throws -> Completion {
        let req = try buildRequest(messages: messages, tools: tools, stream: false)
        let (data, response) = try await session.data(for: req)
        try Self.throwIfHTTPError(response: response, body: data)
        return try Self.decodeNonStreaming(data)
    }

    public func stream(
        messages: [Message],
        tools: [ToolDeclaration]
    ) -> AsyncThrowingStream<Delta, Error> {
        AsyncThrowingStream { continuation in
            let task = Task {
                do {
                    let req = try self.buildRequest(messages: messages, tools: tools, stream: true)
                    let (bytes, response) = try await self.session.bytes(for: req)
                    try Self.throwIfHTTPError(response: response, body: Data())

                    // Streaming tool_call accumulator — upstream sends
                    // argument fragments across many chunks with a shared
                    // index / id.
                    var liveCalls: [Int: (id: String, name: String)] = [:]

                    for try await line in bytes.lines {
                        if Task.isCancelled { break }
                        // SSE frames look like `data: {...}` with empty
                        // lines between events. We only care about the
                        // JSON payload lines.
                        guard line.hasPrefix("data: ") else { continue }
                        let payload = String(line.dropFirst("data: ".count))
                        if payload == "[DONE]" { break }
                        guard let data = payload.data(using: .utf8) else { continue }
                        let chunk = try JSONDecoder().decode(StreamChunk.self, from: data)
                        guard let choice = chunk.choices.first else { continue }
                        if let text = choice.delta.content, !text.isEmpty {
                            continuation.yield(.text(text))
                        }
                        for tc in choice.delta.tool_calls ?? [] {
                            if let name = tc.function?.name {
                                // First fragment for this index — emit
                                // toolCallStart and remember the id.
                                let id = tc.id ?? "tc_\(tc.index)"
                                liveCalls[tc.index] = (id: id, name: name)
                                continuation.yield(.toolCallStart(id: id, name: name))
                            }
                            if let args = tc.function?.arguments,
                               let (id, _) = liveCalls[tc.index] {
                                continuation.yield(.toolCallArgs(id: id, chunk: args))
                            }
                        }
                    }
                    continuation.yield(.end)
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }

    public func close() { /* URLSession is process-shared; nothing to release. */ }

    // MARK: – Request builder

    private func buildRequest(
        messages: [Message],
        tools: [ToolDeclaration],
        stream: Bool
    ) throws -> URLRequest {
        let url = baseURL.appendingPathComponent("chat/completions")
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let apiKey, !apiKey.isEmpty {
            req.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        }
        if stream {
            req.setValue("text/event-stream", forHTTPHeaderField: "Accept")
        } else {
            req.setValue("application/json", forHTTPHeaderField: "Accept")
        }
        for (k, v) in extraHeaders {
            req.setValue(v, forHTTPHeaderField: k)
        }
        req.httpBody = try Self.encodeBody(
            model: model,
            messages: messages,
            tools: tools,
            stream: stream,
            temperature: temperature,
            maxTokens: maxTokens
        )
        return req
    }

    // MARK: – Encoding / decoding

    private static func encodeBody(
        model: String,
        messages: [Message],
        tools: [ToolDeclaration],
        stream: Bool,
        temperature: Double?,
        maxTokens: Int?
    ) throws -> Data {
        var body: [String: Any] = [
            "model": model,
            "stream": stream,
            "messages": messages.map(wireMessage),
        ]
        if let temperature { body["temperature"] = temperature }
        if let maxTokens { body["max_tokens"] = maxTokens }
        if !tools.isEmpty {
            body["tools"] = tools.map(wireTool)
            body["tool_choice"] = "auto"
        }
        return try JSONSerialization.data(withJSONObject: body, options: [])
    }

    /// Map one `Message` to the OpenAI wire shape. The assistant turn
    /// that CARRIES tool calls populates `tool_calls` with parsed JSON
    /// objects; tool responses use `role=tool` + `tool_call_id`.
    private static func wireMessage(_ m: Message) -> [String: Any] {
        var out: [String: Any] = ["role": m.role.rawValue]
        out["content"] = m.content
        if !m.toolCalls.isEmpty {
            out["tool_calls"] = m.toolCalls.map { tc -> [String: Any] in
                [
                    "id": tc.id,
                    "type": "function",
                    "function": [
                        "name": tc.name,
                        // Arguments are a raw JSON string in our model
                        // and OpenAI wants a JSON string too — pass as-is.
                        "arguments": tc.arguments,
                    ],
                ]
            }
        }
        if let id = m.toolCallId { out["tool_call_id"] = id }
        return out
    }

    private static func wireTool(_ d: ToolDeclaration) -> [String: Any] {
        // JsonSchema.serialize() already returns the OpenAI-compatible
        // JSON Schema for an object type, so we round-trip it through
        // JSONSerialization to embed the resulting dictionary inline.
        let schemaJson = d.parameters.serialize().data(using: .utf8) ?? Data("{}".utf8)
        let schemaObj  = (try? JSONSerialization.jsonObject(with: schemaJson)) ?? [:]
        return [
            "type": "function",
            "function": [
                "name": d.name,
                "description": d.description,
                "parameters": schemaObj,
            ],
        ]
    }

    private static func decodeNonStreaming(_ data: Data) throws -> Completion {
        let reply = try JSONDecoder().decode(NonStreamingResponse.self, from: data)
        guard let choice = reply.choices.first else {
            throw OpenAICompatibleError.emptyResponse
        }
        let message = choice.message
        if let calls = message.tool_calls, !calls.isEmpty {
            let mapped = calls.map { c in
                ToolCall(
                    id: c.id,
                    name: c.function?.name ?? "",
                    arguments: c.function?.arguments ?? "{}"
                )
            }
            return .toolCalls(Message(role: .assistant, content: "", toolCalls: mapped))
        }
        return .text(Message(role: .assistant, content: message.content ?? ""))
    }

    private static func throwIfHTTPError(response: URLResponse, body: Data) throws {
        guard let http = response as? HTTPURLResponse else { return }
        if (200..<300).contains(http.statusCode) { return }
        let preview = String(data: body, encoding: .utf8) ?? "<binary>"
        throw OpenAICompatibleError.httpError(status: http.statusCode, body: preview)
    }
}

// MARK: – Wire types

/// Non-streaming reply: one `choices[0].message` carries either plain
/// text in `content` or `tool_calls` (an array of function invocations).
private struct NonStreamingResponse: Decodable {
    let choices: [Choice]
    struct Choice: Decodable {
        let message: MessageBody
    }
    struct MessageBody: Decodable {
        let content: String?
        let tool_calls: [ToolCallBody]?
    }
    struct ToolCallBody: Decodable {
        let id: String
        let function: FunctionBody?
    }
    struct FunctionBody: Decodable {
        let name: String?
        let arguments: String?
    }
}

/// One SSE streaming chunk. Each chunk carries delta-only payload;
/// concatenate `choices[0].delta.content` across chunks to rebuild
/// the final text, and merge `choices[0].delta.tool_calls[i]` by
/// `index` to assemble the argument JSON.
private struct StreamChunk: Decodable {
    let choices: [StreamChoice]
    struct StreamChoice: Decodable {
        let delta: Delta
    }
    struct Delta: Decodable {
        let content: String?
        let tool_calls: [ToolCallFragment]?
    }
    struct ToolCallFragment: Decodable {
        let index: Int
        let id: String?
        let function: FunctionFragment?
    }
    struct FunctionFragment: Decodable {
        let name: String?
        let arguments: String?
    }
}

// MARK: – Error

public enum OpenAICompatibleError: Error, LocalizedError {
    case httpError(status: Int, body: String)
    case emptyResponse

    public var errorDescription: String? {
        switch self {
        case .httpError(let s, let b): return "OpenAI-compatible HTTP \(s): \(b)"
        case .emptyResponse:           return "OpenAI-compatible response had no choices"
        }
    }
}
